src/main/java/com/thebois/models/world/structures/AbstractStructure.java

```diff
@@ -14,114 +15,84 @@
         this.structureType = structureType;
+        this.allNeededItems = allNeededItems;
     }
 
     /**
      * Creates a structure with a position and type.
      *
      * @param position      The position the structure have.
      * @param structureType The type of structure to create.
+     * @param allNeededItems The items needed to finalize construction.
      */
-    AbstractStructure(Position position, StructureType structureType) {
-        this((int) position.getPosX(), (int) position.getPosY(), structureType);
+    AbstractStructure(
+        Position position,
+        StructureType structureType,
+        final IInventory allNeededItems) {
+        this((int) position.getPosX(), (int) position.getPosY(), structureType, allNeededItems);
     }
 
     @Override
     public Position getPosition() {
         return position.deepClone();
     }
 
-    @Override
-    public boolean equals(final Object o) {
-        if (this == o) {
-            return true;
-        }
-        if (o == null || getClass() != o.getClass()) {
-            return false;
-        }
-        final AbstractStructure that = (AbstractStructure) o;
-        return position.equals(that.position) && structureType == that.structureType;
-    }
-
-    @Override
-    public int hashCode() {
-        return Objects.hash(position, structureType);
-    }
-
     @Override
     public StructureType getType() {
         return structureType;
     }
 
     @Override
-    public Collection<IItem> neededItems() {
-        // Make a copy of all IItems in collectionOfAllNeededItems
-        final Collection<IItem> itemDifference = new ArrayList<>(allNeededItems);
-        // Remove IItems that have been delivered
-        deliveredItems.forEach(itemDifference::remove);
-        // Return difference of the two collections
-        return itemDifference;
+    public Collection<ItemType> getNeededItems() {
+        return allNeededItems.calculateDifference(deliveredItems);
     }
 
     @Override
-    public boolean deliverItem(final IItem deliveredItem) {
-        if (neededItems().contains(deliveredItem)) {
+    public boolean tryDeliverItem(final IItem deliveredItem) {
+        if (getNeededItems().contains(deliveredItem.getType())) {
             deliveredItems.add(deliveredItem);
             return true;
         }
-        else {
-            return false;
-        }
+        return false;
     }
 
     @Override
-    public float builtStatus() {
-        if (this.allNeededItems.isEmpty()) {
-            return 1f;
-        }
-        else {
-            final float totalDelivered = this.deliveredItems.size();
-            final float totalNeeded = this.allNeededItems.size();
+    public float getBuiltRatio() {
+        final float totalDelivered = this.deliveredItems.size();
+        final float totalNeeded = this.allNeededItems.size();
 
-            return totalDelivered / totalNeeded;
-        }
+        return totalDelivered / totalNeeded;
     }
```

Comment:
Are you assuming that neededItems is never empty? If so, there should probably a check for that in the constructor so that if someone tries to provide an empty list of needed items, it throws an InvalidArgumentException.